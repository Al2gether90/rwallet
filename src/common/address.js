class Address {
  static isSameAddr(addr1, addr2) {
    const keys = ['address', 'type', 'chain', 'symbol'];
    return keys.every((key) => addr1[key] === addr2[key]);
  }
}

export default Address;
