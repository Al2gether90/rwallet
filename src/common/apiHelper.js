import axios from 'axios';
import Rsk3 from '@rsksmart/rsk3';

import ParseHelper from './parse';

const headers = {
  'X-Parse-Application-Id': 'rwallet',
  'X-Parse-REST-API-KEY': '',
  'Content-Type': 'application/json',
  'X-RWALLET-API-KEY': 'masterKey',
};

const apiHelper = {
  async getAbiByAddress(address) {
    try {
      const parseServer = ParseHelper.getServerUrl();
      const url = `${parseServer}/functions/getAbiByAddress`;
      // using Rsk3.utils.toChecksumAddress without chainId to match the backend to get the erc20 token's info(symbol, abi)
      const result = await axios.post(url, { address: Rsk3.utils.toChecksumAddress(address) }, { headers, timeout: 5000 });
      return result.data.result;
    } catch (error) {
      console.log('getAbiByAddress error: ', error);
      return null;
    }
  },
};

export default apiHelper;
