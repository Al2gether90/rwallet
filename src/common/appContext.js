import I18n from 'react-native-i18n';
import RNSecureStorage from 'rn-secure-storage';
import storage from './storage';
import ParseHelper from './parse';

const appContext = {
  loadData() {
    this.walletKeyId = 0;
  },
  data: {
    walletKeyId: 0,
    wallets: [],
    language: I18n.currentLocale(),
    user: null,
    settings: {
      fingerprint: false,
    },
  },
  user: null,
  async set(key, value) {
    this.data[key] = value;
    await storage.save('data', this.data);
  },
  secureSet(key, value, callback = null) {
    RNSecureStorage.set(key, value, {}).then((res) => {
      console.log(`RNSecureStorage.set, res: ${res}`);
      if (callback) {
        callback();
      }
    }, (err) => {
      console.log(err);
    });
  },
  secureGet(key, callback) {
    RNSecureStorage.exists(key).then((res) => {
      if (!res) {
        callback(null);
      } else {
        RNSecureStorage.get(key).then((value) => {
          console.log(`RNSecureStorage.get, key: ${key}`);
          if (callback) {
            callback(value);
          }
        }, (err) => {
          console.log(err);
        });
      }
    }, (err) => {
      console.log(err);
    });
  },
  async saveSettings(settings) {
    Object.assign(this.data.settings, settings);
    await this.set('settings', this.data.settings);
    await ParseHelper.uploadSettings(this.user, this.data.settings);
  },
  async saveWallets(wallets) {
    await this.set('wallets', wallets);
    const uploadWallets = [];
    wallets.forEach((wallet) => {
      const item = { coins: wallet.coins };
      uploadWallets.push(item);
    });
    await ParseHelper.uploadWallets(this.user, uploadWallets);
  },
};

export default appContext;
