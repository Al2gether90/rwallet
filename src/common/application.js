import _ from 'lodash';
import { Map } from 'immutable';
import UUIDGenerator from 'react-native-uuid-generator';
import storage from './storage';

class Application {
  constructor() {
    this.data = Map({
      id: undefined,
    });

    this.deserialize = this.deserialize.bind(this);
  }

  get(key) {
    return this.data.get(key);
  }

  set(key, value) {
    this.data = this.data.set(key, value);
  }

  toJSON() {
    return this.data.toJSON();
  }

  /**
   * Save JSON presentation of settings data to permenate storage
   */
  async serialize() {
    const jsonData = this.data.toJSON();
    await storage.save('app', jsonData);
  }

  /**
   * Read permenate storage and load settins into this instance;
   * Load default settings if Settings is empty in storage
   */
  async deserialize() {
    const result = await storage.load({ key: 'app' });

    console.log('Deserialized Application from Storage.', result);

    if (!_.isNull(result)) {
      this.data = Map(result);
      return;
    }

    // If result is null; meaning there's no Application in storage
    // therefore we create a new one here
    const newId = await this.createId();
    await this.saveId(newId);
  }

  createId = async () => new Promise((resolve) => {
    UUIDGenerator.getRandomUUID((uuid) => resolve(uuid));
  })

  saveId = async (id) => {
    this.set('id', id);
    await this.serialize();
  }
}

export default new Application();
