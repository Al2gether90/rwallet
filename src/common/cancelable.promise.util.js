import _ from 'lodash';

/**
 * Usage:
 *
 * Use CancelablePromiseUtil in page
 * Wrap promise with CancelablePromiseUtil.makeCancelable(promise, this);
 *
 * const getRatePromise = CancelablePromiseUtil.makeCancelable(CoinswitchHelper.getRate(...), this);
 * const sdRate = await getRatePromise.promise;
 *
 * Cancel all the promise about the page when componentWillUnmount
 *
 * componentWillUnmount() {
 *   CancelablePromiseUtil.cancel(this);
 * }
 */
const CancelablePromiseUtil = {
  promiseObjects: [],

  // Wrap promise, return the promise wrapper { tag, promise, cancel }
  makeCancelable(promise, tag) {
    let hasCanceled = false;

    const wrappedPromise = new Promise((resolve, reject) => {
      promise.then(
        (val) => {
          if (!hasCanceled) {
            resolve(val);
          }
          _.remove(this.promiseObjects, (promiseObject) => promiseObject.promise === wrappedPromise);
        },
        (error) => {
          if (hasCanceled) {
            reject(error);
          }
          _.remove(this.promiseObjects, (promiseObject) => promiseObject.promise === wrappedPromise);
        },
      );
    });

    const promiseObject = {
      tag,
      promise: wrappedPromise,
      cancel() {
        hasCanceled = true;
      },
    };

    this.promiseObjects.push(promiseObject);
    return promiseObject;
  },

  // cancel with tag
  cancel(tag) {
    _.each(this.promiseObjects, (promiseObject) => {
      if (promiseObject.tag === tag) {
        promiseObject.cancel();
      }
    });
    _.remove(this.promiseObjects, { tag });
    console.log(this.promiseObjects);
  },
};

export default CancelablePromiseUtil;
