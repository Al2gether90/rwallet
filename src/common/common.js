import React from 'react';
import { Text, Platform } from 'react-native';
import BigNumber from 'bignumber.js';
import _ from 'lodash';
import { Toast } from '@ant-design/react-native';
import * as bitcoin from 'bitcoinjs-lib';
import rsk3 from 'rsk3';
import config from '../../config';
import store from './storage';


const { consts: { currencies } } = config;
const DEFAULT_CURRENCY_SYMBOL = currencies[0].symbol;

// Extract currency symbols from config
// Generate {USD: '$', RMB: '￥', ARS: 'ARS$', KRW: '₩', JPY: '￥', GBP: '£',}
const currencySymbols = _.reduce(currencies, (obj, row) => {
  const settingsObj = obj;
  settingsObj[row.name] = row.symbol;
  return settingsObj;
}, {});

const common = {
  currentNavigation: null,
  isIphoneX() {
    // TODO
    // return DeviceInfo.getModel().toLowerCase().indexOf('iphone x') >= 0
    return false;
  },
  btcToSatoshiHex(amount) {
    const result = `0x${new BigNumber(amount).times('1e8').decimalPlaces(0).toString(16)}`;
    return result;
  },
  satoshiToBtc(satoshi) {
    const result = new BigNumber(satoshi).div('1e8');
    return result;
  },
  rbtcToWeiHex(amount) {
    const result = `0x${new BigNumber(amount).times('1e18').decimalPlaces(0).toString(16)}`;
    return result;
  },
  rbtcToWei(amount) {
    const result = new BigNumber(amount).times('1e18');
    return result;
  },
  weiToRbtc(wei) {
    const result = new BigNumber(wei).div('1e18');
    return result;
  },
  rifToWeiHex(amount) {
    const result = `0x${new BigNumber(amount).times('1e18').decimalPlaces(0).toString(16)}`;
    return result;
  },
  weiToRif(wei) {
    const result = new BigNumber(wei).div('1e18');
    return result;
  },
  Toast(text, type, onClose, duration, mask) {
    const last = duration > 0 ? duration : 1.5;
    if (type === 'success') {
      Toast.success(text, last, onClose, mask);
    } else if (type === 'fail') {
      Toast.fail(text, last, onClose, mask);
    } else { // none
      Toast.info(text, last, onClose, mask);
    }
  },
  convertUnitToCoinAmount(symbol, unitNumber) {
    let amount = null;
    switch (symbol) {
      case 'BTC':
        amount = common.satoshiToBtc(unitNumber);
        break;
      case 'RBTC':
        amount = common.weiToRbtc(unitNumber);
        break;
      case 'RIF':
        amount = common.weiToRif(unitNumber);
        break;
      default:
    }
    return amount;
  },
  /**
   * getBalanceString, diffrent symbol apply diffrent decimalPlaces, subfix 0 will be omitted.
   * The balance will be round down by default.
   * @param {string} symbol
   * @param {BigNumber | number | string} balance
   */
  getBalanceString(symbol, balance) {
    const decimalPlaces = config.symbolDecimalPlaces[symbol];
    if (!_.isNull(balance)) {
      let balanceBigNumber = balance;
      if (typeof balance === 'number' || typeof value === 'string') {
        balanceBigNumber = new BigNumber(balance);
      }
      return balanceBigNumber.decimalPlaces(decimalPlaces, BigNumber.ROUND_DOWN).toFixed();
    }
    return null;
  },
  /**
   * getAssetValueString, value apply default decimalPlaces, subfix 0 will be omitted.
   * @param {BigNumber | number | string} value
   */
  getAssetValueString(value) {
    if (!_.isNull(value)) {
      let valueBigNumber = value;
      if (typeof value === 'number' || typeof value === 'string') {
        valueBigNumber = new BigNumber(value);
      }
      return valueBigNumber.decimalPlaces(config.assetValueDecimalPlaces).toFixed();
    }
    return null;
  },
  getCoinPrice(symbol, currency, prices) {
    for (let i = 0; i < prices.length; i += 1) {
      const priceRow = prices[i];
      if (symbol === priceRow.symbol) {
        const price = priceRow.price[currency];
        return price;
      }
    }
    return null;
  },
  getCoinValue(amount, symbol, currency, prices) {
    if (!amount || !prices || prices.length === 0) {
      return null;
    }
    try {
      const price = this.getCoinPrice(symbol, currency, prices);
      const amountBigNumber = new BigNumber(amount);
      const value = amountBigNumber.times(price);
      return value;
    } catch (e) {
      console.error(e);
    }
    return null;
  },

  getCurrencyNames() {
    return _.map(currencies, (item) => item.name);
  },

  /**
   * Get currency symbol string for example '$' based on currency
  * @param {string} currency currency string such as 'USD'
  */
  getCurrencySymbol(currency) {
    if (currencySymbols[currency]) {
      return currencySymbols[currency];
    }

    return DEFAULT_CURRENCY_SYMBOL;
  },

  async updateInAppPasscode(input) {
    let passcode = null;
    if (input) {
      await store.setPasscode(input);
      // eslint-disable-next-line no-multi-assign
      global.passcode = passcode = input;
    } else {
      // eslint-disable-next-line no-multi-assign
      global.passcode = passcode = await store.getPasscode();
    }
    return passcode;
  },

  /**
   * getTransactionUrl, returns transaction url
   * @param {*} symbol, coin symbol
   * @param {*} type, coin network type
   * @param {*} hash, transaction hash
   */
  getTransactionUrl(symbol, type, hash) {
    let url = '';
    if (config.transactionUrls[symbol] && config.transactionUrls[symbol][type]) {
      url = config.transactionUrls[symbol][type];
    }
    url = `${url}/${hash}/`;
    return url;
  },

  /**
   * getLatestBlockHeight, return latestBlockHeight. If it's not found, return null.
   * @param {array} latestBlockHeights
   * @param {string} chain
   * @param {string} type, network type
   */
  getLatestBlockHeight(latestBlockHeights, chain, type) {
    const latestBlockHeight = _.find(latestBlockHeights, { chain, type });
    if (latestBlockHeight && latestBlockHeight.blockHeight) {
      return latestBlockHeight.blockHeight;
    }
    return null;
  },

  /**
   * Validate btc address
   * @param {string} address
   * @param {string} type, MainTest or Testnet
   */
  isBtcAddress(address, type) {
    // https://github.com/bitcoinjs/bitcoinjs-lib/issues/890
    // https://bitcoin.stackexchange.com/questions/52740/how-do-you-validate-a-bitcoin-address-using-bitcoinjs-library-in-javascript
    try {
      let network = null;
      if (type === 'Testnet') {
        network = bitcoin.networks.testnet;
      }
      bitcoin.address.toOutputScript(address, network);
      return true;
    } catch (e) {
      return false;
    }
  },

  /**
   * Validate wallet address
   * @param {string} address
   * @param {string} symbol, BTC, RBTC, RIF
   * @param {string} type, MainTest or Testnet
   */
  isWalletAddress(address, symbol, type) {
    let isAdress = false;
    if (symbol === 'BTC') {
      isAdress = this.isBtcAddress(address, type);
    } else {
      const checksumAddress = rsk3.utils.toChecksumAddress(address);
      isAdress = rsk3.utils.isAddress(checksumAddress);
    }
    return isAdress;
  },

  /**
   * Validate amount
   * @param {string} str
   */
  isAmount(str) {
    const regex = /^\d*\.{0,1}\d+$/g;
    return regex.test(str);
  },

  /**
   * Set default font family for android, solve cut-off problem for some android device
   * Oppo A77 - Some texts gets cut-off
   * solution: Set app default font family, instead of system font
   * see https://github.com/facebook/react-native/issues/15114
   */
  setDefaultFontFamily() {
    if (Platform.OS !== 'android') {
      return;
    }

    const oldRender = Text.render;
    Text.render = (...args) => {
      const origin = oldRender.call(this, ...args);
      return React.cloneElement(origin, {
        style: [{ fontFamily: config.defaultFontFamily }, origin.props.style],
      });
    };
  },

  /**
   * getSymbolFullName
   * @param {string} symbol, BTC, RBTC, RIF
   * @param {string} type, MainTest or Testnet
   */
  getSymbolFullName(symbol, type) {
    return `${type === 'Testnet' ? 'Test' : ''} ${symbol}`;
  },
};

export default common;
