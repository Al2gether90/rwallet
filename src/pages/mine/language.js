import React, { Component } from 'react';
import {
  View, StyleSheet,
} from 'react-native';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Header from '../../components/common/misc/header';
import flex from '../../assets/styles/layout.flex';
import SelectionList from '../../components/common/list/selectionList';
import actions from '../../redux/languageSwitcher/actions';

const styles = StyleSheet.create({
  buttonView: {
    position: 'absolute',
    bottom: '5%',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
  },
  listView: {
    width: '80%',
    alignSelf: 'center',
  },
});

class Language extends Component {
    static navigationOptions = () => ({
      header: null,
    });

    listData = [
      {
        title: 'English',
      },
      {
        title: 'French',
      },
      {
        title: 'Hebrew',
      },
      {
        title: 'Chinese',
      },
    ];

    constructor(props) {
      super(props);
      this.onChange = this.onChange.bind(this);
    }

    onChange(index) {
      const { changeLanguage } = this.props;
      const locales = ['en', 'fr', 'he', 'zh'];
      changeLanguage(locales[index]);
    }

    render() {
      const { language } = this.props;
      const selected = {
        en: 0, fr: 1, he: 2, zh: 3,
      }[language];
      const { navigation } = this.props;
      const goBack = navigation && navigation.goBack ? navigation.goBack : () => {};
      return (
        <View style={[flex.flex1]}>
          <Header title="Language" goBack={goBack} />
          <View style={styles.listView}>
            <SelectionList data={this.listData} onChange={this.onChange} selected={selected} />
          </View>
        </View>
      );
    }
}

Language.propTypes = {
  navigation: PropTypes.shape({
    navigate: PropTypes.func.isRequired,
    dispatch: PropTypes.func.isRequired,
    goBack: PropTypes.func.isRequired,
    state: PropTypes.object.isRequired,
  }).isRequired,
  changeLanguage: PropTypes.func.isRequired,
  language: PropTypes.string.isRequired,
};

const mapStateToProps = (state) => ({
  language: state.LanguageSwitcher.get('language'),
});

const mapDispatchToProps = (dispatch) => ({
  changeLanguage: (language) => dispatch(
    actions.changeLanguage(language),
  ),
});

export {
  Language,
};
export default connect(mapStateToProps, mapDispatchToProps)(Language);
