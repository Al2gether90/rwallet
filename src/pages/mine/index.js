import React, { Component } from 'react';
import {
  View, Text, StyleSheet, Image, TouchableOpacity, ScrollView, FlatList, ImageBackground, Linking,
} from 'react-native';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import Entypo from 'react-native-vector-icons/Entypo';
import MaterialIcons from 'react-native-vector-icons/MaterialIcons';
import FontAwesome from 'react-native-vector-icons/FontAwesome';
import FontAwesome5 from 'react-native-vector-icons/FontAwesome5';
import Loc from '../../components/common/misc/loc';
import flex from '../../assets/styles/layout.flex';
import { DEVICE } from '../../common/info';
import { strings } from '../../common/i18n';
import ScreenHelper from '../../common/screenHelper';
import presetStyles from '../../assets/styles/style';

const styles = StyleSheet.create({
  sectionTitle: {
    marginTop: 5,
    fontSize: 20,
    fontWeight: '600',
    color: '#000',
    marginBottom: 20,
  },
  sectionContainer: {
    paddingHorizontal: 30,
  },
  body: {
    paddingTop: 65,
  },
  avatar: {
    position: 'absolute',
    left: 20,
    bottom: -40,
  },
  row: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingLeft: 10,
  },
  title: {
    color: '#0B0B0B',
    fontSize: 16,
  },
  right: {
    flexDirection: 'row',
    alignItems: 'center',
    marginLeft: 15,
    borderBottomWidth: 1,
    borderBottomColor: '#EDEDED',
    paddingVertical: 20,
    flex: 1,
  },
  logoView: {
    alignItems: 'center',
    flexDirection: 'row',
    position: 'absolute',
    bottom: 0,
    right: 0,
  },
  powerby: {
    color: '#727372',
    fontSize: 17,
    fontWeight: '900',
    marginTop: 5,
  },
  nameView: {
    alignItems: 'center',
    flexDirection: 'row',
    position: 'absolute',
    left: 160,
    bottom: 5,
  },
  name: {
    fontSize: 20,
    fontWeight: '900',
    letterSpacing: 0.39,
    color: '#FFFFFF',
    paddingBottom: 6,
  },
  nameEditView: {
    padding: 5,
  },
  nameEdit: {
    color: '#FFFFFF',
    marginLeft: 10,
  },
  communityIcon: {
    marginLeft: -5.5,
  },
  keyListView: {
    marginLeft: 10,
  },
  keyListRow: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  keyListRowRight: {
    alignItems: 'center',
  },
  keyIcon: {
    color: '#4A4A4A', transform: [{ rotate: '90deg' }, { rotateX: '180deg' }],
  },
  keyTitle: {
    marginLeft: 10,
  },
  keyWallets: {
    backgroundColor: '#F3F3F3',
    borderRadius: 5,
    padding: 5,
    color: '#000',
    position: 'absolute',
    right: 0,
  },
  createWalletButtonView: {
    marginTop: 20,
    marginBottom: 5,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    padding: 5,
  },
  createWalletButtonText: {
    color: '#00B520',
    fontSize: 16,
  },
  lastBlockMarginBottom: {
    marginBottom: 15,
  },
});

const header = require('../../assets/images/misc/header.png');
const avatar = require('../../assets/images/mine/avatar.png');
const rsk = require('../../assets/images/mine/rsk.png');

function Item({ data, title }) {
  return (
    <TouchableOpacity
      style={[styles.row]}
      onPress={() => {
        if (data.onPress) {
          data.onPress();
        }
      }}
    >
      {data.icon}
      <View style={styles.right}>
        <Text>{title}</Text>
      </View>
    </TouchableOpacity>
  );
}

Item.propTypes = {
  data: PropTypes.shape({
    title: PropTypes.string.isRequired,
    icon: PropTypes.element.isRequired,
    onPress: PropTypes.func,
  }),
  title: PropTypes.string.isRequired,
};

Item.defaultProps = {
  data: { onPress: null },
};


const ICON_SIZE = 20;

class MineIndex extends Component {
  static navigationOptions = () => ({
    header: null,
  });

  static createKeyListData(wallets) {
    const listData = [];
    wallets.forEach((wallet) => {
      const item = {
        name: wallet.name,
        walletCount: wallet.coins.length,
        wallet,
      };
      listData.push(item);
    });
    return listData;
  }

  static renderKeyListView(listData, navigation) {
    return (
      <FlatList
        style={styles.keyListView}
        data={listData}
        renderItem={({ item }) => (
          <TouchableOpacity
            style={styles.keyListRow}
            onPress={() => navigation.navigate('KeySettings', { key: item.wallet })}
          >
            <FontAwesome5 name="key" size={20} style={styles.keyIcon} />
            <View style={styles.right}>
              <Text>{item.name}</Text>
              <Text style={styles.keyWallets}>
                {`${item.walletCount} `}
                <Loc text="Wallets" />
              </Text>
            </View>
          </TouchableOpacity>
        )}
        keyExtractor={(item, index) => index.toString()}
      />
    );
  }

  settings = [
    {
      title: 'Language',
      icon: <MaterialIcons name="language" size={ICON_SIZE} style={{ color: '#4A4A4A' }} />,
      onPress: () => {
        const { navigation } = this.props;
        navigation.navigate('Language');
      },
    },
    {
      title: 'Currency',
      icon: (
        <MaterialCommunityIcons name="currency-usd" size={ICON_SIZE} style={{ color: '#4A4A4A' }} />
      ),
      onPress: () => {
        const { navigation } = this.props;
        navigation.navigate('Currency');
      },
    },
    // {
    //   title: 'Two-Factor Authentication',
    //   icon: (
    //     <MaterialCommunityIcons
    //       name="two-factor-authentication"
    //       size={ICON_SIZE}
    //       style={{ color: '#4A4A4A' }}
    //     />
    //   ),
    //   onPress: () => {
    //     const { navigation } = this.props;
    //     navigation.navigate('TwoFactorAuth');
    //   },
    // },
  ];

  joins = [
    {
      title: 'Twitter',
      icon: <FontAwesome name="twitter" size={30} style={[styles.communityIcon, { color: '#039BE5' }]} />,
      onPress: () => {
        Linking.openURL('https://twitter.com/rsksmart');
      },
    },
    {
      title: 'Telegram',
      icon: <FontAwesome name="telegram" size={30} style={[styles.communityIcon, { color: '#6FC062' }]} />,
      onPress: () => {
        Linking.openURL('https://t.me/rskofficialcommunity');
      },
    },
    {
      title: 'Facebook',
      icon: <Entypo name="facebook-with-circle" size={30} style={[styles.communityIcon, { color: '#3F51B5' }]} />,
      onPress: () => {
        Linking.openURL('https://www.facebook.com/RSKsmart/');
      },
    },
    {
      title: 'Gitter',
      icon: <FontAwesome5 name="gitter" size={30} style={[styles.communityIcon]} />,
      onPress: () => {
        Linking.openURL('https://gitter.im/rsksmart');
      },
    },
    {
      title: 'Reddit',
      icon: <FontAwesome name="reddit" size={30} style={[styles.communityIcon, { color: '#FF4500' }]} />,
      onPress: () => {
        Linking.openURL('https://www.reddit.com/r/rootstock/');
      },
    },
    {
      title: 'YouTube',
      icon: <Entypo name="youtube-with-circle" size={30} style={[styles.communityIcon, { color: '#D2142B' }]} />,
      onPress: () => {
        Linking.openURL('https://www.youtube.com/rsksmart');
      },
    },
  ];

  constructor(props) {
    super(props);
    this.state = {
      keyListData: [],
      settings: [],
      joins: [],
    };
  }

  componentWillMount() {
    const { wallets } = this.props;
    const keyListData = MineIndex.createKeyListData(wallets);
    this.setState({
      keyListData,
      settings: this.settings,
      joins: this.joins,
    });
  }

  componentWillReceiveProps(nextProps) {
    const { isWalletsUpdated, isWalletNameUpdated, wallets } = nextProps;
    if ((isWalletsUpdated || isWalletNameUpdated) && wallets) {
      const keyListData = MineIndex.createKeyListData(wallets);
      this.setState({ keyListData });
    }
  }

  render() {
    let headerHeight = 160;
    if (DEVICE.isIphoneX) {
      headerHeight += ScreenHelper.iphoneXExtendedHeight;
    }
    const { language, navigation } = this.props;
    const { keyListData, settings, joins } = this.state;
    return (
      <View style={flex.flex1}>
        <ScrollView style={[flex.flex1]}>
          <ImageBackground source={header} style={[{ height: headerHeight }]}>
            <Image source={avatar} style={styles.avatar} />
            <View style={styles.nameView}>
              <Text style={styles.name}>Jean Payene</Text>
              <TouchableOpacity style={styles.nameEditView}>
                <FontAwesome name="edit" size={25} style={styles.nameEdit} />
              </TouchableOpacity>
            </View>
          </ImageBackground>
          <View style={[styles.body]}>
            <View style={styles.sectionContainer}>
              <Loc style={[styles.sectionTitle]} text="Settings" />
              <FlatList
                data={settings}
                extraData={language}
                renderItem={({ item }) => <Item data={item} title={strings(item.title)} />}
                keyExtractor={(item, index) => index.toString()}
              />
            </View>
          </View>
          <View style={[styles.sectionContainer, { marginTop: 10 }]}>
            <Loc style={[styles.sectionTitle]} text="Keys" />
            {MineIndex.renderKeyListView(keyListData, navigation)}
            <View style={styles.createWalletButtonView}>
              <TouchableOpacity onPress={() => navigation.navigate('WalletAddIndex')}>
                <Loc style={[styles.createWalletButtonText]} text="Create or Import a Key" />
              </TouchableOpacity>
            </View>
          </View>
          <View style={[styles.sectionContainer, styles.lastBlockMarginBottom, { marginTop: 10 }]}>
            <Loc style={[styles.sectionTitle]} text="Join RSK's community" />
            <FlatList
              data={joins}
              renderItem={({ item }) => <Item data={item} title={item.title} />}
              keyExtractor={(item, index) => index.toString()}
            />
          </View>
        </ScrollView>
        <View style={styles.logoView}>
          <Loc style={[styles.powerby]} text="Powered by" />
          <Image style={presetStyles.rskIcon} source={rsk} />
        </View>
      </View>
    );
  }
}

MineIndex.propTypes = {
  navigation: PropTypes.shape({
    navigate: PropTypes.func.isRequired,
    dispatch: PropTypes.func.isRequired,
    goBack: PropTypes.func.isRequired,
    state: PropTypes.object.isRequired,
  }).isRequired,
  isWalletsUpdated: PropTypes.bool.isRequired,
  isWalletNameUpdated: PropTypes.bool.isRequired,
  language: PropTypes.string.isRequired,
  wallets: PropTypes.arrayOf(PropTypes.object),
};

MineIndex.defaultProps = {
  wallets: undefined,
};

const mapStateToProps = (state) => ({
  language: state.App.get('language'),
  wallets: state.Wallet.get('walletManager') && state.Wallet.get('walletManager').wallets,
  isWalletsUpdated: state.Wallet.get('isWalletsUpdated'),
  isWalletNameUpdated: state.Wallet.get('isWalletNameUpdated'),
});

export default connect(mapStateToProps)(MineIndex);
