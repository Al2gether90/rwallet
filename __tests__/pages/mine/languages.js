import React from 'react';
import renderer from 'react-test-renderer';

import { mockComponent } from '../../../src/common/testUtils'

import {Language} from'../../../src/pages/mine/language';
jest.mock('../../../src/components/common/misc/header', () => { return mockComponent('Icon') });

describe("Button component", () => {
    test('renders correctly', () => {
        const tree = renderer.create(<Language />).toJSON();
        expect(tree).toMatchSnapshot();
    });
    test('onchange correctly', () => {
        const component = renderer.create(<Language changeLanguage={() => {}}/>);
        const instance = component.getInstance();
        expect(instance.state).toBe(null);
    })
});
