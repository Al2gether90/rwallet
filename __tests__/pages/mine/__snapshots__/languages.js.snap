// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
    ]
  }
>
  <mocked
    goBack={[Function]}
    originalComponent="Icon"
    title="Language"
  />
  <View
    style={
      Object {
        "alignSelf": "center",
        "width": "80%",
      }
    }
  >
    <RCTScrollView
      data={
        Array [
          Object {
            "title": "English",
          },
          Object {
            "title": "French",
          },
          Object {
            "title": "Hebrew",
          },
          Object {
            "title": "Chinese",
          },
        ]
      }
      disableVirtualization={false}
      extraData={
        Object {
          "selectedIndex": 0,
        }
      }
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "borderBottomColor": "#bbb",
                    "borderBottomWidth": 0.5,
                    "flexDirection": "row",
                    "paddingVertical": 30,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#2D2D2D",
                        "flex": 1,
                        "fontSize": 16,
                      },
                    ]
                  }
                >
                  English
                </Text>
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": undefined,
                        "fontSize": 20,
                      },
                      Object {
                        "color": "#00B520",
                        "marginRight": 10,
                      },
                      Object {
                        "fontFamily": "anticon",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  î˜²
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "borderBottomColor": "#bbb",
                    "borderBottomWidth": 0.5,
                    "flexDirection": "row",
                    "paddingVertical": 30,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#2D2D2D",
                        "flex": 1,
                        "fontSize": 16,
                      },
                    ]
                  }
                >
                  French
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "borderBottomColor": "#bbb",
                    "borderBottomWidth": 0.5,
                    "flexDirection": "row",
                    "paddingVertical": 30,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#2D2D2D",
                        "flex": 1,
                        "fontSize": 16,
                      },
                    ]
                  }
                >
                  Hebrew
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "borderBottomColor": "#bbb",
                    "borderBottomWidth": 0.5,
                    "flexDirection": "row",
                    "paddingVertical": 30,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#2D2D2D",
                        "flex": 1,
                        "fontSize": 16,
                      },
                    ]
                  }
                >
                  Chinese
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
